ml-agents/mlagents/trainers/components/reward_signals/gail/model.py

```diff
@@ -228,7 +231,35 @@ def create_network(self) -> None:
         self.discriminator_score = tf.reshape(
             self.policy_estimate, [-1], name="GAIL_reward"
         )
-        self.intrinsic_reward = -tf.log(1.0 - self.discriminator_score + 1e-7)
+        self.intrinsic_reward = -tf.log(1.0 - self.discriminator_score + EPSILON)
+
+    def compute_gradient_penalty(self) -> tf.Tensor:
+        """
+        Gradient penalty from https://arxiv.org/pdf/1704.00028. Adds stability esp.
+        for off-policy. Compute gradients w.r.t randomly interpolated input.
+        """
+        expert = [self.encoded_expert, self.expert_action, self.done_expert]
+        policy = [
+            self.encoded_policy,
+            self.policy_model.selected_actions,
+            self.done_policy,
+        ]
+        interp = []
+        for _expert_in, _policy_in in zip(expert, policy):
+            alpha = tf.random_uniform(tf.shape(_expert_in))
+            interp.append(alpha * _expert_in + (1 - alpha) * _policy_in)
+
+        grad_estimate, _, grad_input = self.create_encoder(
+            interp[0], interp[1], interp[2], reuse=True
+        )
+
+        grad = tf.gradients(grad_estimate, [grad_input])[0]
+
+        # Norm, like log, can return NaN. Use our own safe_norm
+        safe_norm = tf.sqrt(tf.reduce_sum(grad ** 2, axis=-1) + EPSILON)
```

Comment:
How does norm result in NaN? I could see that happening if there was overflow, but in that case adding an epsilon isn't going to help.

Reply:
Not the norm, it's the gradient of the norm. At 0 the gradient of sqrt() is a horizontal line. I'll update the comment to reflect this.